File: modules/swagger-core/src/test/scala/1_3/converter/XMLGregorianCalendarTest.scala
package converter

import javax.xml.datatype.XMLGregorianCalendar

import com.fasterxml.jackson.module.scala.DefaultScalaModule
import io.swagger.annotations.ApiModelProperty
import io.swagger.converter.ModelConverters
import io.swagger.models.properties.{DateTimeProperty, StringProperty}
import io.swagger.util.Json
import org.junit.runner.RunWith
import org.scalatest.{FlatSpec, Matchers}
import org.scalatest.junit.JUnitRunner

import scala.annotation.meta.field

@RunWith(classOf[JUnitRunner])
class XMLGregorianCalendarTest extends FlatSpec with Matchers {
  Json.mapper().registerModule(DefaultScalaModule)

  it should "read a model with XMLGregorianCalendar" in {
    val models = ModelConverters.getInstance().readAll(classOf[ModelWithCalendar])
    models.size should be(1) // don't create a Joda DateTime object

    val model = models.get("ModelWithCalendar")
    val properties = model.getProperties()

    val nameProperty = properties.get("name")
    nameProperty.isInstanceOf[StringProperty] should be(true)
    nameProperty.getPosition() should be(2)
    nameProperty.getDescription() should be("name of the model")

    val dateTimeProperty = properties.get("createdAt")
    dateTimeProperty.isInstanceOf[DateTimeProperty] should be(true)
    dateTimeProperty.getPosition() should be(1)
    dateTimeProperty.getRequired() should be(true)
    dateTimeProperty.getDescription() should be("creation timestamp")
  }
}


case class ModelWithCalendar(
                              @(ApiModelProperty@field)(value = "name of the model", position = 2) name: String,
                              @(ApiModelProperty@field)(value = "creation timestamp", required = true, position = 1) createdAt: XMLGregorianCalendar)

File: modules/swagger-core/src/test/java/io/swagger/XMLGregorianCalendarTest.java
package io.swagger;

import io.swagger.converter.ModelConverters;
import io.swagger.oas.models.media.DateTimeSchema;
import io.swagger.oas.models.media.Schema;
import io.swagger.oas.models.media.StringSchema;
import org.testng.annotations.Test;

import javax.xml.datatype.XMLGregorianCalendar;
import java.util.Map;

import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertTrue;

public class XMLGregorianCalendarTest {
    @Test(description = "it should read a model with XMLGregorianCalendar")
    public void testXMLGregorianCalendar() {
        final Map<String, Schema> models = ModelConverters.getInstance().readAll(ModelWithCalendar.class);
        assertEquals(models.size(), 1); // don't create a Joda DateTime object

        Schema model = models.get("ModelWithCalendar");

        final Map<String, Schema> properties = model.getProperties();

        final Schema nameProperty = properties.get("name");
        assertTrue(nameProperty instanceof StringSchema);
        assertEquals(nameProperty.getDescription(), "name of the model");

        final Schema dateTimeSchema = properties.get("createdAt");
        assertTrue(dateTimeSchema instanceof DateTimeSchema);
        assertTrue(model.getRequired().contains("createdAt"));
        assertEquals(dateTimeSchema.getDescription(), "creation timestamp");
    }

    class ModelWithCalendar {
        @io.swagger.oas.annotations.media.Schema(description = "name of the model")
        public String name;

        @io.swagger.oas.annotations.media.Schema(description = "creation timestamp", required = true)
        public XMLGregorianCalendar createdAt;
    }
}
